### ISSUE TITLE
BLE connection problem with STM32WB5MMG

### ISSUE THREAD

**Issue by LoWnT13** [stm32duino/STM32duinoBLE#56]
Hi,

I just bought a ST32WB5MMG module from ST and I tried to use a simple BLE code but I can't get a BLE connection with any device.
I already tried to flash and start "stm32wb5x_BLE_HCILayer_fw.bin" at address "0x080E0000" following the Release Notes for STM32WB Copro Wireless Binaries but it didn't help.
I ask for help to know if there is a special procedure to activate it.

Thank you for your help.
LT

**Comment by fpistm** [comment]
Hi @LoWnT13 
when you told `ST32WB5MMG module` could you be more precise? which target you select in the Arudino IDE menu?

**Comment by LoWnT13** [comment]
I use:  Generic STM32WB series --> Generic WB5MMGHx

**Comment by fpistm** [comment]
OK. So example used specific board definition:
https://github.com/stm32duino/STM32duinoBLE/blob/cf77ef0349afcec3cdd8ec231405d41670645ac2/examples/Peripheral/ButtonLED/ButtonLED.ino#L60

So you have to add the generic one and probably have to configure `SystemClock_Config` to use HSE.

**Comment by fpistm** [comment]
I've tested the Generic variant with my STM32WBMM-DK.
I've made these change to the ButtonLed.ino examples:

1- Redefine the clock config:
```

#include "lock_resource.h"

extern "C" void SystemClock_Config(void) {
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};

  /* This prevents concurrent access to RCC registers by CPU2 (M0+) */
  hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);

  /** Configure LSE Drive Capability
  */
  HAL_PWR_EnableBkUpAccess();
  __HAL_RCC_LSEDRIVE_CONFIG(RCC_LSEDRIVE_LOW);
  /** Configure the main internal regulator output voltage
  */
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);

  /* This prevents the CPU2 (M0+) to disable the HSI48 oscillator */
  hsem_lock(CFG_HW_CLK48_CONFIG_SEMID, HSEM_LOCK_DEFAULT_RETRY);

  /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI48 | RCC_OSCILLATORTYPE_HSI
                                     | RCC_OSCILLATORTYPE_HSE | RCC_OSCILLATORTYPE_LSE;
  RCC_OscInitStruct.HSEState = RCC_HSE_ON;
  RCC_OscInitStruct.LSEState = RCC_LSE_ON;
  RCC_OscInitStruct.HSIState = RCC_HSI_ON;
  RCC_OscInitStruct.HSI48State = RCC_HSI48_ON;
  RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
  RCC_OscInitStruct.PLL.PLLM = RCC_PLLM_DIV2;
  RCC_OscInitStruct.PLL.PLLN = 8;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLR = RCC_PLLR_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = RCC_PLLQ_DIV2;
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
    Error_Handler();
  }
  /** Configure the SYSCLKSource, HCLK, PCLK1 and PCLK2 clocks dividers
  */
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK4 | RCC_CLOCKTYPE_HCLK2
                                | RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
                                | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.AHBCLK2Divider = RCC_SYSCLK_DIV2;
  RCC_ClkInitStruct.AHBCLK4Divider = RCC_SYSCLK_DIV1;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_3) != HAL_OK) {
    Error_Handler();
  }
  /** Initializes the peripherals clocks
  */
  /* RNG needs to be configured like in M0 core, i.e. with HSI48 */
  PeriphClkInitStruct.PeriphClockSelection = RCC_PERIPHCLK_SMPS | RCC_PERIPHCLK_RFWAKEUP
                                             | RCC_PERIPHCLK_CLK48SEL | RCC_PERIPHCLK_USB
                                             | RCC_PERIPHCLK_RNG;
  PeriphClkInitStruct.UsbClockSelection = RCC_USBCLKSOURCE_HSI48;
  PeriphClkInitStruct.RngClockSelection = RCC_RNGCLKSOURCE_HSI48;
  PeriphClkInitStruct.RFWakeUpClockSelection = RCC_RFWKPCLKSOURCE_LSE;
  PeriphClkInitStruct.SmpsClockSelection = RCC_SMPSCLKSOURCE_HSE;
  PeriphClkInitStruct.SmpsDivSelection = RCC_SMPSCLKDIV_RANGE0;
  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
    Error_Handler();
  }

  LL_PWR_SMPS_SetStartupCurrent(LL_PWR_SMPS_STARTUP_CURRENT_80MA);
  LL_PWR_SMPS_SetOutputVoltageLevel(LL_PWR_SMPS_OUTPUT_VOLTAGE_1V40);
  LL_PWR_SMPS_Enable();

  /* Select HSI as system clock source after Wake Up from Stop mode */
  LL_RCC_SetClkAfterWakeFromStop(LL_RCC_STOP_WAKEUPCLOCK_HSI);

  hsem_unlock(CFG_HW_RCC_SEMID);
}
```

2- Then change the `LED_BUILTIN `and `USER_BTN `to use as the generic define them as `PNUM_NOT_DEFINED`
```
const int ledPin = PA2;
const int buttonPin = PC12;  // set buttonPin to on-board user button
```

3- Change default `Serial `pins as generic define on by default but probably does not match to your board:
```
  Serial.setRx(PB7);
  Serial.setTx(PB6);
  Serial.begin(115200);
```

4- Add board definition to define the `HCISharedMemTransportClass HCISharedMemTransport;`

`#elif defined(ARDUINO_NUCLEO_WB15CC) || defined(ARDUINO_P_NUCLEO_WB55RG) || defined(ARDUINO_STM32WB5MM_DK) || defined(P_NUCLEO_WB55_USB_DONGLE) || defined(ARDUINO_GENERIC_WB5MMGHX)`

Then it works as expected.
So you have to properly update the example to fit your board.

**Comment by fpistm** [comment]
Hi @jmchiappa 
I've checked and the fw was an old one. I've updated and it does not work.
After some investigation, the fw was not properly updated. Updating it a second time and it works.
Think to press start wireless stack too in the Cube Programmer.
![image](https://user-images.githubusercontent.com/20641798/235869909-f57017c7-a844-4245-a9d4-695f2aa14ecb.png)

![image](https://user-images.githubusercontent.com/20641798/235870387-11442dd0-3e55-4808-933e-7e0ac5b780a7.png)
